clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -1130,6 +1130,13 @@ invariant_checker_t::check_for_pc_discontinuity(
          (shard->prev_xfer_marker_.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT ||
           shard->prev_xfer_marker_.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_XFER ||
           shard->prev_xfer_marker_.marker.marker_type == TRACE_MARKER_TYPE_RSEQ_ABORT)) ||
+#ifdef UNIX
+        // Same PC is allowed for back-to-back signals without any intervening
+        // instructions.
+        (cur_pc == shard->last_signal_context_.xfer_int_pc &&
+         prev_instr_trace_pc ==
+             shard->last_signal_context_.pre_signal_instr.memref.instr.addr) ||
```

Comment:
Trying to wrap my head around this: so we already allow the pre-signal-instr PC to equal the post-signal PC (for synchronous signal), up at line 1127?  Oh, but that's just on interruption: the check on sigreturn is done elsewhere at line 642.  So line 1127 is allowing the pre-signal-instr PC to equal the interruption PC.  But that's not enough here b/c the rseq abort causes the pre-signal-instr PC to not equal the interruption PC which is the rseq handler.

So a key thing here is that last_signal_context_ is *not* yet updated: it holds the prior signal, not the one we're entering.  Please add a comment about that.  (Otherwise these checks don't make sense.). With that fact: ok, yes, it does make sense.

Reply:
Done.